lib/Sema/CSApply.cpp

```diff
@@ -5985,9 +5985,60 @@ Expr *ExprRewriter::coerceToType(Expr *expr, Type toType,
                                  toTuple->getElementForScalarInit(), locator);
     }
 
-    case ConversionRestrictionKind::DeepEquality:
-      assert(toType->hasUnresolvedType() && "Should have handled this above");
-      break;
+    case ConversionRestrictionKind::DeepEquality: {
+      if (toType->hasUnresolvedType())
+        break;
+
+      // HACK: Fix problem related to Swift 3 mode (with assertions),
+      // since Swift 3 mode allows passing arguments with extra parens
+      // to parameters which don't expect them, it should be supported
+      // by "deep equality" types like Optional<T> e.g.
+      // ```swift
+      // func foo(_: (() -> Void)?) {}
+      // func bar() -> ((()) -> Void)? { return nil }
+      // foo(bar) // This expression should compile in Swift 3 mode
+      // ```
+      if (tc.getLangOpts().isSwiftVersion3()) {
+        auto bound1 = fromType->getAs<BoundGenericType>();
+        auto bound2 = toType->getAs<BoundGenericType>();
+
+        if (!bound1 || !bound2)
+          llvm_unreachable("Should be handled above");
+
+        auto genericArgs1 = bound1->getGenericArgs();
+        auto genericArgs2 = bound2->getGenericArgs();
```

Comment:
How about using `getAnyOptionalObjectType()` here, because this is regarding optionals? It'll make for a much smaller change.

Reply:
I don't think it's only options, it's any generic type, but I can make it only about optionals...

Reply:
Huh. Okay!

Reply:
If it's more general that optionals, then leave it as is. Thank you for clarifying!

Reply:
E.g.

```swift
struct S<T> {}
let x: S<(()) -> Void> = S()
func foo(_: S<() -> Void>) {}
foo(x) // should work in 3 mode
```

Reply:
I'll add one more test like this just for clarity and merge, thank you!

Reply:
@slavapestov @DougGregor Looks like it doesn't work if I simply break in case of `S<() -> Void>` vs. `S<(()) -> Void>` it goes through whole `coerceToType` method and fails on unreachable as unhandled coercion (if I `return expr` it fails later on saying that types didn't match up in ShuffleExpr). Do you know if there is any specific coercion that could be applied here or should make it so the hack works only for optionals as @DougGregor suggested and let everything else fall-through into unreachable?

Reply:
I guess if there is no coercion like that we shouldn't even be forming solutions in Swift 3 mode...

Reply:
@DougGregor it looks like we can't simply fix Verifier to ignore this because it requires function-to-function conversion but I can see that we only support conversion + injection for optionals and not other bound generic types, is that correct? I think the best would be to make it so only optional produces solution in swift 3 and other bound generic types do not, WDYT?